src/data/loaders/geojson.ts

```diff
@@ -0,0 +1,198 @@
+    const { signal } = options;
+    signal?.throwIfAborted();
+    const geo = await this._loadGeoJSON({ signal });
+    signal?.throwIfAborted();
+    const multipolygons = this._parseGeoJSONToMultiPolygons(geo);
+    signal?.throwIfAborted();
+    return new GeoJSONShapesData(multipolygons);
+  }
+
+  private async _loadGeoJSON(
+    options: { signal?: AbortSignal } = {},
+  ): Promise<GeoJSONObject> {
+    const { signal } = options;
+    signal?.throwIfAborted();
+    if (this.dataSource.path !== undefined && this.workspace !== null) {
+      const fh = await this.workspace.getFileHandle(this.dataSource.path);
+      signal?.throwIfAborted();
+      const file = await fh.getFile();
+      signal?.throwIfAborted();
+      const text = await file.text();
+      return JSON.parse(text) as GeoJSONObject;
+    }
+    if (this.dataSource.url !== undefined) {
+      const resp = await fetch(this.dataSource.url);
+      if (!resp.ok) {
+        throw new Error(
+          `Failed to fetch GeoJSON: ${resp.status} ${resp.statusText}`,
+        );
+      }
+      const text = await resp.text();
+      return JSON.parse(text) as GeoJSONObject;
+    }
+    if (this.dataSource.path !== undefined) {
+      throw new Error("An open workspace is required to open local-only data.");
+    }
+    throw new Error("A URL or workspace path is required to load data.");
+  }
+
+  private _parseGeoJSONToMultiPolygons(obj: GeoJSONObject): MultiPolygon[] {
+    if (!obj) return [];
+
+    const multiPolygons: MultiPolygon[] = [];
+
+    const pushFromGeometry = (geometry: GeoJSONGeometry | null) => {
+      if (!geometry || !("type" in geometry)) return;
+      const type = geometry.type;
+      if (type === "Polygon") {
+        const coords = (geometry as GeoJSONPolygon).coordinates;
+        const polygon = this._coordsToPolygon(coords);
+        multiPolygons.push({ polygons: [polygon] });
+      } else if (type === "MultiPolygon") {
+        const coords = (geometry as GeoJSONMultiPolygon).coordinates;
+        const polygons = coords.map((polygonCoords) =>
+          this._coordsToPolygon(polygonCoords),
+        );
+        multiPolygons.push({ polygons });
+      }
+    };
+
+    // FeatureCollection
+    if ((obj as GeoJSONFeatureCollection).type === "FeatureCollection") {
+      const fc = obj as GeoJSONFeatureCollection;
+      for (const feature of fc.features) {
+        if (!feature || !feature.geometry) continue;
+        pushFromGeometry(feature.geometry);
+      }
+    } else if ((obj as GeoJSONFeature).type === "Feature") {
+      const f = obj as GeoJSONFeature;
+      pushFromGeometry(f.geometry);
+    } else if (
+      (obj as GeoJSONGeometry).type === "Polygon" ||
+      (obj as GeoJSONGeometry).type === "MultiPolygon"
+    ) {
+      pushFromGeometry(obj as GeoJSONGeometry);
+    }
+
+    return multiPolygons;
+  }
+
+  private _coordsToPolygon(
+    coords: GeoJSONPosition[][],
+  ): MultiPolygon["polygons"][0] {
```

Comment:
What's `coords`? Shouldn't the return type be just `Polygon`? In general, I think this function needs some TLC